File: router/src/harness/src/stdx/filesystem.cc
/*
Copyright (c) 2020, 2022, Oracle and/or its affiliates.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License, version 2.0,
as published by the Free Software Foundation.

This program is also distributed with certain software (including
but not limited to OpenSSL) that is licensed under separate terms,
as designated in a particular file or component or in included license
documentation.  The authors of MySQL hereby grant you an additional
permission to link the program and your derivative works with the
separately licensed software that they have included with MySQL.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mysql/harness/stdx/filesystem.h"

#ifndef _WIN32
#include <unistd.h>  // unlink
#else
#include <Windows.h>  // GetLastError
#include <direct.h>   // _getcwd, _rmdir
#include <io.h>       // _unlink
#endif

#include <array>
#include <climits>
#include <system_error>

#include "mysql/harness/stdx/expected.h"

#if !defined(PATH_MAX) && defined(MAX_PATH)
#define PATH_MAX MAX_PATH
#endif

namespace stdx {
namespace filesystem {

static std::error_code last_posix_error_code() {
  return {errno, std::generic_category()};
}

static std::error_code last_error_code() {
#ifdef _WIN32
  return {static_cast<int>(GetLastError()), std::system_category()};
#else
  return last_posix_error_code();
#endif
}

namespace impl {
stdx::expected<void, std::error_code> unlink(const char *path_name) {
#ifdef _WIN32
  if (-1 == ::_unlink(path_name)) {
    return stdx::make_unexpected(last_posix_error_code());
  }
#else
  if (-1 == ::unlink(path_name)) {
    return stdx::make_unexpected(last_error_code());
  }
#endif
  return {};
}

stdx::expected<void, std::error_code> rmdir(const char *path_name) {
#ifdef _WIN32
  if (-1 == ::_rmdir(path_name)) {
    return stdx::make_unexpected(last_posix_error_code());
  }
#else
  if (-1 == ::rmdir(path_name)) {
    return stdx::make_unexpected(last_error_code());
  }
#endif
  return {};
}

stdx::expected<std::string, std::error_code> getcwd() {
#ifdef _WIN32
  std::array<char, MAX_PATH> cwd;
  if (nullptr == ::_getcwd(cwd.data(), cwd.size())) {
    // _getcwd() sets errno
    return stdx::make_unexpected(last_posix_error_code());
  }
#else
  std::array<char, PATH_MAX> cwd{};
  if (nullptr == ::getcwd(cwd.data(), cwd.size())) {
    return stdx::make_unexpected(last_error_code());
  }
#endif

  return {cwd.data()};
}

}  // namespace impl

path temp_directory_path(std::error_code &ec) {
  ec.clear();
#ifdef _WIN32
  std::array<char, MAX_PATH + 1> d;
  auto sz = GetTempPath(d.size(), d.data());
  if (sz == 0) {
    ec = last_error_code();
    return {};
  }

  return {std::string(d.begin(), std::next(d.begin(), sz))};
#else
  for (auto const *envvar : {"TMPDIR", "TMP", "TEMP", "TEMPDIR"}) {
    auto *path = getenv(envvar);
    if (path) return {path};
  }

  return {"/tmp"};
#endif
}

path current_path(std::error_code &ec) noexcept {
  ec.clear();

  auto res = impl::getcwd();
  if (res) return res.value();

  ec = res.error();

  return {};
}

path current_path() {
  std::error_code ec;

  auto p = current_path(ec);
  if (ec) throw std::system_error(ec);

  return p;
}

bool remove(const path &p, std::error_code &ec) noexcept {
  ec.clear();

  auto res = impl::unlink(p.c_str());

  if (!res && res.error() == make_error_condition(std::errc::is_a_directory)) {
    res = impl::rmdir(p.c_str());
  }

  if (res) return true;

  ec = res.error();

  return false;
}

}  // namespace filesystem
}  // namespace stdx

File: libs/scripts/impl/file_system.py
#!/usr/bin/python3.11
# Copyright (c) 2024, Oracle and/or its affiliates.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2.0,
# as published by the Free Software Foundation.
#
# This program is designed to work with certain software (including
# but not limited to OpenSSL) that is licensed under separate terms,
# as designated in a particular file or component or in included license
# documentation.  The authors of MySQL hereby grant you an additional
# permission to link the program and your derivative works with the
# separately licensed software that they have either included with
# the program or referenced in the documentation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License, version 2.0, for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA

from   debug import trace
from   messages import log_error, log_info

import os
from   typing import Iterator

'''
Helper functions for working with the file system.
'''

@trace
def walk_tree(*,
        root: str,
        recurse: bool=True,
        yield_directories: bool=True,
        yield_root: bool=None,
        topdown: bool=True
) -> Iterator[str]:
    '''Iterator over files within the given directory.

    This computes the set of files once and then traverses them. Thus, it is
    reliable even if the user moves the files.

    Within a given directory, yields files before subdirectories.

    Yields filenames relative to *root*.

    Parameters
    ----------
    :param root: The root directory
    :param recurse: Recurse into subdirectories.
    :param yield_directories: If True, yield both files and directories within
        the traversed tree. If False, filter out subdirectories (but still
        recurse into them, if recurse==True.)
    :param yield_root: If True, yield the root of the tree. If False, don't.
        Defaults to the same value as *yield_directories*.
    :param topdown: If True, yield parents before children. If False, yield
        children before parents. Contrary to os.walk, modifying the yielded
        items has no impact on the traversal, even when topdown=True.
    '''

    if yield_root is None:
        yield_root = yield_directories
    # Pre-compute the complete list, so that it is stable
    # even if the directory structure changes while we
    # traverse it
    iterator = os.walk(root, topdown)
    if recurse:
        full_list = list(iterator)
    else:
        first = next(iterator, None)
        full_list = [first] if first else []
    for dirpath, dirnames, filenames in full_list:
        relpath = os.path.relpath(dirpath, root)
        if relpath == '.':
            relpath = ''
            if yield_root and topdown:
                yield ''
        for filename in filenames:
            yield os.path.join(relpath, filename)
        if yield_directories:
            for dirname in dirnames:
                yield os.path.join(relpath, dirname)
        if relpath == '' and yield_root and not topdown:
            yield ''

def is_empty_dir(path: str) -> bool:
    '''Return true if *path* is a directory without contents.'''

    return not next(walk_tree(root=path, yield_root=False), None)

@trace
def makedirs(path: str) -> None:
    '''Create all missing trailing path components of *path*.'''

    prefix = path
    while not os.path.exists(prefix):
        prefix = os.path.dirname(prefix)
    if not os.path.isdir(prefix):
        log_error(
            f'Cannot create "{path}" because "{prefix}" is a non-directory.'
        )
    if prefix == path: return
    path_to_create = os.path.relpath(path, prefix)
    log_info(f'Creating {path_to_create} under {prefix}')
    os.makedirs(path)

@trace
def prune_empty_dirs(path) -> str:
    '''Prune remove path if it is empty, then do the same for the parent.

    Return the deepest of the parent directories that still exists.
    '''

    while path and is_empty_dir(path):
        log_info(f'Pruning {path}')
        os.rmdir(path)
        path = os.path.dirname(path)
    return path

@trace
def rename(*, old: str, new: str) -> None:
    '''Rename *old* file to *new*, pruning empty parent directories for old.

    Return the deepest of the parent directories that still exists.
    '''

    log_info(f'Renaming {old} to {new}')
    os.rename(old, new)
    return prune_empty_dirs(os.path.dirname(old))

@trace
def remove_file(file: str) -> None:
    '''Remove *file*, pruning empty parent directories.

    Returns the deepest of the parent directory that still exists.
    '''

    log_info(f'Removing {file}')
    os.remove(file)
    return prune_empty_dirs(os.path.dirname(file))
